src/Transport/GrpcTransport.php

```diff
@@ -283,4 +307,66 @@ private static function loadClientCertSource(callable $clientCertSource)
     {
         return call_user_func($clientCertSource);
     }
+
+    /**
+     * @param Call $call
+     * @param array $headers
+     *
+     * @return void
+     */
+    private function logRequest(Call $call, array $headers)
+    {
+        $logger = $this->logger;
+        $timestamp = date(DATE_RFC3339);
+
+        $debugEvent = [
+            'timestamp' => $timestamp,
+            'severity' => 'DEBUG',
+            'jsonPayload' => [
+                'request.rpcName' => $call->getMethod(),
+                'request.headers' => $headers,
+                'request.payload' => $call->getMessage()
+            ]
+        ];
+
+        $logger->debug(json_encode($debugEvent));
+    }
+
+    /**
+     * @param mixed $response
+     * @param mixed $status
+     *
+     * @return void
+     */
+    private function logResponse(mixed $response, mixed $status)
+    {
+        $logger = $this->logger;
+        $timestamp = date(DATE_RFC3339);
+
+        // In the case we have a $status->code != Code::OK
+        // from the request, we log the status only and we do not have
+        // a response.
+        if ($response) {
+            $debugEvent = [
+                'timestamp' => $timestamp,
+                'severity' => 'DEBUG',
+                'jsonPayload' => [
+                    'response.headers' => $status->metadata,
+                    'response.payload' => $response,
```

Comment:
If the `$response` is of type `mixed`, it makes me nervous just shoving it into `json_encode` like this. Can we not narrow it down more than this? For instance, if it's a `ResponseInterface`, we may want to log `(string) $response->getBody()`.

Additionally, if this is a protobuf `Message` class, then `json_encode` just shows us `{}`:

```php
$message = new Google\Cloud\Logging\V2\Link(); // any message
$message->setName('Link Name');
var_dump(json_encode($message));
// string(2) "{}"
```

But if we use `serializeToJsonString`, we get to see the data:

```php
$message = new Google\Cloud\Logging\V2\Link(); // any message
$message->setName('Link Name');
var_dump(json_encode($message->serializeToJsonString()));
// string(16) "{"name":"brent"}"
```

So, I think we want to do more handling with this object than just putting everything in `json_encode`

Reply:
I think `Message` would be appropriate, then. Looking at that code, it assumes a `mergeFromString` function exists. So it COULD be a generic object (e.g. duck typing), but I think we can assume it's a `Message`.  

From the [link you provided](https://github.com/grpc/grpc/blob/master/src/php/lib/Grpc/AbstractCall.php#L127), at the very least it must be an `object` or `null`. So I'd recommend doing something like this (to be safe): 

```php
private function logResponse(object|null $response, mixed $status)
{
    // .. 
    if ($response) {
        $responseToSerialize = $response instanceof Message
            ? $response->serializeToJsonString()
            : get_class($response);
    // ..
```

That way in some case where it isn't a `Message` (which we don't expect, but which is technically possible), we can at least handle that case gracefully.